cdap-app-fabric/src/main/java/co/cask/cdap/metadata/FieldLineageAdmin.java

```
@@ -0,0 +1,322 @@
+      List<InputField> inputFields = new ArrayList<>();
+      switch (operation.getType()) {
+        case TRANSFORM:
+          TransformOperation transform = (TransformOperation) operation;
+          inputFields.addAll(transform.getInputs());
+          break;
+        case WRITE:
+          WriteOperation write = (WriteOperation) operation;
+          inputFields.addAll(write.getInputs());
+          outgoingOperations.put(operation.getName(), new HashSet<>());
+          break;
+      }
+
+      for (InputField inputField : inputFields) {
+        // Add current operation to the adjacency list of operation represented by origin
+        Set<Operation> outgoings = outgoingOperations.computeIfAbsent(inputField.getOrigin(), k -> new HashSet<>());
+        outgoings.add(operation);
+      }
+    }
+
+    Stack<Operation> stack = new Stack<>();
+    Set<String> visitedOperations = new HashSet<>();
+    for (Operation currentOperation : operations) {
+      String currentOperationName = currentOperation.getName();
+      if (visitedOperations.contains(currentOperationName)) {
+        continue;
+      }
+      topologicalSortUtil(outgoingOperations, currentOperation, visitedOperations, stack);
+    }
+
+    List<Operation> topologicalOrder = new ArrayList<>();
+    while (!stack.empty()) {
+      topologicalOrder.add(stack.pop());
+    }
+    return topologicalOrder;
+  }
+
+  /**
+   * Helper function for performing recursion while finding the topological sorting.
+   */
+  private void topologicalSortUtil(Map<String, Set<Operation>> outgoingOperations, Operation currentOperation,
+                                   Set<String> visitedOperations, Stack<Operation> stack) {
+    String currentOperationName = currentOperation.getName();
+    // Mark the current operation as visited.
+    visitedOperations.add(currentOperationName);
+
+    Set<Operation> outgoings = outgoingOperations.get(currentOperationName);
+    if (outgoings == null) {
```

When is it null? Seems like this shouldn't happen unless there is a malformed map passed in?

It seems odd that this is treated differently than if outgoings is an empty set. Here it will not get pushed onto the stack, with an empty set it will. We should either throw an exception here if it is never expected, or push the current operation onto the stack before returning.

---

It wont be null, so no need to do the null check.